src/cliffordsynthesis/encoding/PhaseCorrectionEncoder.cpp

```diff
@@ -0,0 +1,148 @@
+  DEBUG() << "Encoding pauli constraints.";
+  splitXorR(initialPhase);
+  for (std::size_t q = 0U; q < N; ++q) {
+    const auto& xCol = vectorFromBitset(uncorrected.getBVFrom(q));
+    const auto& zCol = vectorFromBitset(uncorrected.getBVFrom(N + q));
+
+    LogicVector iChange{};
+    LogicVector xChange{};
+    LogicVector zChange{};
+    LogicVector yChange{};
+    iChange.reserve(S);
+    xChange.reserve(S);
+    zChange.reserve(S);
+    yChange.reserve(S);
+
+    for (std::size_t row = 0U; row < S; ++row) {
+      DEBUG() << "Encoding identity constraint.";
+      iChange.emplace_back(logicbase::LogicTerm::ite(
+          paulis[q][0], LogicTerm(false), LogicTerm(false)));
+      DEBUG() << "Encoding X constraint.";
+      xChange.emplace_back(
+          logicbase::LogicTerm::ite(paulis[q][1], zCol[row], LogicTerm(false)));
+      DEBUG() << "Encoding Y constraint.";
+      yChange.emplace_back(logicbase::LogicTerm::ite(
+          paulis[q][2], !(zCol[row] == xCol[row]), LogicTerm(false)));
+      DEBUG() << "Encoding Z constraint.";
+      zChange.emplace_back(
+          logicbase::LogicTerm::ite(paulis[q][3], xCol[row], LogicTerm(false)));
+    }
+    splitXorR(iChange);
+    splitXorR(xChange);
+    splitXorR(yChange);
+    splitXorR(zChange);
+
+    // at least one gate per qubit
+    lb->assertFormula(paulis[q][0] || paulis[q][1] || paulis[q][2] ||
+                      paulis[q][3]);
+  }
+  DEBUG() << "Assert Phase.";
+  for (std::size_t row = 0U; row < S; ++row) {
+    lb->assertFormula(targetPhase[row] == xorHelpers.back()[row]);
+  }
+  DEBUG() << "Finished Phase Encoding.";
+}
+
+void PhaseCorrectionEncoder::splitXorR(const LogicVector& changes) {
+  auto xorHelper = LogicVector{};
+  xorHelper.reserve(S);
+  for (std::size_t row = 0U; row < S; ++row) {
+    const std::string hName =
+        "h_" + std::to_string(row) + "_" + std::to_string(xorHelpers.size());
+    DEBUG() << "Creating helper variable for RChange XOR " << hName;
+    xorHelper.emplace_back(lb->makeVariable(hName));
+  }
+  xorHelpers.emplace_back(xorHelper);
+  if (xorHelpers.size() == 1) {
+    for (std::size_t row = 0U; row < S; ++row) {
+      lb->assertFormula(xorHelpers[0][row] == changes[row]);
+    }
+  } else {
+    for (std::size_t row = 0U; row < S; ++row) {
+      lb->assertFormula(
+          xorHelpers.back()[row] ==
+          (changes[row] != xorHelpers[xorHelpers.size() - 2][row]));
+    }
+  }
+}
+
+LogicVector
+PhaseCorrectionEncoder::vectorFromBitset(const std::bitset<64>& bs) const {
+  LogicVector result{};
+  result.reserve(S);
+  for (std::size_t i = 0U; i < S; ++i) {
+    result.emplace_back(bs[i]);
+  }
+  return result;
+}
+
+std::vector<qc::OpType> PhaseCorrectionEncoder::extractResult() {
+  const auto&             model = lb->getModel();
+  std::vector<qc::OpType> result{};
+  result.reserve(N);
+  for (std::size_t q = 0U; q < N; ++q) {
+    GateSet pauliGates{};
+    pauliGates.removePaulis();
```

Comment:
This is a truly strange sequence of calls to me that only makes me ask: WHY? 🙃 
You create an empty gate set and then you remove Paulis from that?